pipic/doc/pipic.1
.\" Automatically generated by Pod::Man 2.22 (Pod::Simple 3.13)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "PIPIC 1"
.TH PIPIC 1 "2013-09-09" "version 20130909" "Raspberry Pi"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
pipic \-  send commands and read data from PiPIC
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBpipic\fR \fB\-a\fR i2c address [\fB\-c\fR command [\fB\-d\fR data]] [\fB\-r\fR b|w|W] 
[\fB\-h\fR] [\fB\-v\fR] [\fB\-V\fR]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The \fBpipic\fR is used to send commands to a \s-1PIC\s0 processor on i2c bus connected
to Raspberry Pi. The command is one byte of data and is given with option
\&\fB\-c\fR as a hexadecimal 0x00 \- 0xFF. The command byte can be followed by
one to four bytes of parameter data. The parameter is given with switch
\&\fB\-d\fR as a decimal integer with 'b', 'w' or 'W' at end. The \s-1PIC\s0 processor 
has the necessary command interpretation loop programmed in. After the 
command has been executed the possible results can be read with option \fB\-r\fR
by specifying the length of data 'b' for byte, 'w' for 16\-bit word and
\&'W' for 32\-bit word.
.SH "OPTIONS"
.IX Header "OPTIONS"
\&\fB\-a\fR chip address on i2c bus
.PP
\&\fB\-c\fR send command byte to \s-1PIC\s0
.PP
\&\fB\-d\fR data byte (b), 16\-bit word (w) or 32\-bit word (W) after command byte
.PP
\&\fB\-r\fR read byte (b), 16\-bit word (w) or 32\-bit word (W) from \s-1PIC\s0
.PP
\&\fB\-h\fR display a short help text
.PP
\&\fB\-v\fR verbose
.PP
\&\fB\-V\fR print version
.SH "COMMANDS"
.IX Header "COMMANDS"
0x00 do nothing
.PP
0x01 0xNN read byte from file address and write it to output buffer
.PP
0x02 0xNNNNNNNN write four bytes to transmit buffer, used for debugging
.PP
0x03 0xNN read \s-1EEPROM\s0 address
.PP
0x04 0xNNNN write to \s-1EEPROM\s0 address
.PP
0x05 read \s-1TMR1\s0 as 16\-bit word
.PP
0x06 0xC5 reinitialize
.PP
0x10 clear GPIO0=0 output
.PP
0x11 clear GPIO1=0 output
.PP
0x14 clear GPIO4=0 output
.PP
0x15 clear GPIO5=0 output
.PP
0x20 set GPIO0=1 output
.PP
0x21 set GPIO1=1 output
.PP
0x24 set GPIO4=1 output
.PP
0x25 set GPIO5=1 output
.PP
0x30 0xNN write byte to \s-1GPIO\s0
.PP
0x31 0xNN write byte to \s-1TRISIO\s0
.PP
0x32 0xNN \s-1AND\s0 \s-1GPIO\s0 with byte
.PP
0x33 0xNN \s-1OR\s0 \s-1GPIO\s0 with byte
.PP
0x34 0xNN \s-1XOR\s0 \s-1GPIO\s0 with byte
.PP
0x40 read analog input \s-1AN0\s0
.PP
0x41 read analog input \s-1AN1\s0
.PP
0x43 read analog input \s-1AN3\s0
.PP
0x50 reset timer
.PP
0x51 read timer
.PP
0x60 stop timed task1
.PP
0x61 start timed task1
.PP
0x62 0x00NNNNNN set task1 counting down time
.PP
0x63 0xNNNN set task1 command and one parameter byte
.PP
0x64 0xNN set how many times task1 is repeated
.PP
0x70 stop timed task2
.PP
0x71 start timed task2
.PP
0x72 0x00NNNNNN set task2 counting down time
.PP
0x73 0xNNNN set task2 command and one parameter byte
.PP
0x74 0xNN set how many times task2 is repeated
.PP
0xA0 disable event triggered tasks
.PP
0xA1 enable event triggered tasks
.PP
0xA2 read event register
.PP
0xA3 reset event register
.PP
0xA4 0xNNNN \s-1GP0\s0 input at zero triggered task command and parameter byte
.PP
0xA5 0xNNNN \s-1GP1\s0 input at zero triggered task command and parameter byte
.PP
0xA6 0xNNNN \s-1GP4\s0 input at zero triggered task command and parameter byte
.PP
0xA7 0xNNNN \s-1GP5\s0 input at zero triggered task command and parameter byte
.PP
0xA8 0xNNNN comparator triggered task command and parameter byte
.SH "CONFIGURATION"
.IX Header "CONFIGURATION"
Configuration with \s-1AN0\s0 on \s-1GP0\s0 for A/D, \s-1GP4\s0 as digital output, \s-1GP5\s0 digital 
input, and timer counting every 0.52 s. First initial value for \s-1CMCON\s0 
register is 0x07 or 0xF8 with bits inverted. The 0x10 is the \s-1EEPROM\s0 memory 
address.
.PP
\&\fBpipicfile\fR \fB\-a\fR 26 \fB\-w\fR 10F8
.PP
The initial value for \s-1GPIO\s0 is 0x00 and this gives 0x11FF.
.PP
Initial value for \s-1ADCON0\s0 is 0x81. This gives 0x127E.
.PP
The \s-1ANSEL\s0 initially is 0x13. This gives 0x13EC.
.PP
The \s-1VRCON\s0 initially is 0x00. This gives 0x14FF.
.PP
The \s-1TRISIO\s0 bits are not inverted. Thus 0x152F.
.PP
For T1CON initial value 0x31 gives 0x16CE.
.PP
The \s-1IOC\s0 initially 0x20 which gives 0x17DF.
.PP
The PiPIC needs to be power cycled before the new configuration is set.
The other possibility is to send reinitialize command
.PP
\&\fBpipic\fR \fB\-a\fR 0x26 \fB\-c\fR 0x06 \fB\-d\fR 197b
.SH "EXAMPLE"
.IX Header "EXAMPLE"
Test i2c communication
.PP
\&\fBpipic\fR \fB\-a\fR 0x26 \fB\-c\fR 0x02 \fB\-d\fR 1375211866W \fB\-r\fR W \fB\-v\fR
.PP
Set \s-1GPIO4\s0 output to \*(L"1\*(R", chip address 0x26 (or 38) and command 0x24 (or 36)
.PP
\&\fBpipic\fR \fB\-a\fR 0x26 \fB\-c\fR 0x24
.PP
This assumes that \s-1GP4\s0 has been configured as digital output.
.PP
Read status register at address 0x03 from PIC12f675, chip address 0x26 (or 38)
.PP
\&\fBpipic\fR \fB\-a\fR 26 \fB\-c\fR 1 \fB\-d\fR 3b \fB\-r\fR b \fB\-v\fR
.PP
Read analog voltage from \s-1AN0\s0
.PP
\&\fBpipic\fR \fB\-a\fR 26 \fB\-c\fR 40 \fB\-r\fR w
.PP
This assumes that \s-1GP0\s0 has been configured as analog input.
.PP
Reset internal timer
.PP
\&\fBpipic\fR \fB\-a\fR 26 \fB\-c\fR 50
.PP
Read internal timer
.PP
\&\fBpipic\fR \fB\-a\fR 26 \fB\-c\fR 51 \fB\-r\fR W
.PP
Write 'Pi' to \s-1EEPROM\s0. First 'P' to memory location 0x00
.PP
\&\fBpipic\fR \fB\-a\fR 26 \fB\-c\fR 04 \fB\-d\fR 80w
.PP
then 'i' to memory location 0x01 (1*256+105=361)
.PP
\&\fBpipic\fR \fB\-a\fR 26 \fB\-c\fR 04 \fB\-d\fR 361w
.PP
Blink \s-1LED\s0 connected to \s-1GP4\s0 three times. First set delay to 2 timer
cycles (usually one timer cycle is 0.524288 s).
.PP
\&\fBpipic\fR \fB\-a\fR 26 \fB\-c\fR 62 \fB\-d\fR 2W
.PP
The \s-1XOR\s0 command 0x34 together with 0x10 for \s-1GPIO4\s0 gives 13328 in decimal.
.PP
\&\fBpipic\fR \fB\-a\fR 26 \fB\-c\fR 63 \fB\-d\fR 13328w
.PP
Repeate \s-1XOR\s0 command on \s-1GPIO4\s0 six times (programmed as minus one)
.PP
\&\fBpipic\fR \fB\-a\fR 26 \fB\-c\fR 64 \fB\-d\fR 5b
.PP
Start task
.PP
\&\fBpipic\fR \fB\-a\fR 26 \fB\-c\fR 61
.PP
Read event register
.PP
\&\fBpipic\fR \fB\-a\fR 26 \fB\-c\fR A2 \fB\-r\fR b \fB\-v\fR
.PP
Reset event register
.PP
\&\fBpipic\fR \fB\-a\fR 26 \fB\-c\fR A3
.PP
Switch on/off \s-1LED\s0 connected to \s-1GP4\s0. The \s-1XOR\s0 command 0x34 together with 0x10 
for \s-1GPIO4\s0 gives 13328 in decimal. The switch connected to \s-1GP5\s0.
.PP
\&\fBpipic\fR \fB\-a\fR 26 \fB\-c\fR A7 \fB\-d\fR 13328w
.PP
Enable event triggered tasks
.PP
\&\fBpipic\fR \fB\-a\fR 26 \fB\-c\fR A1
.SH "FILES"
.IX Header "FILES"
\&\fI/etc/modprobe.d/i2c.conf\fR
.PP
This file should have a line like
.PP
options i2c_bcm2708 baudrate=20000
.PP
which sets the \s-1SCL\s0 clock to 20 kHz.
.PP
\&\fI/etc/modprobe.d/raspi\-blacklist.conf\fR
.PP
Check that \fIi2c_bcm2708\fR is commented out in this file.
.SH "WARNING"
.IX Header "WARNING"
No checking is done where the data is written. Could make some hardware 
unusable.
.SH "AUTHORS"
.IX Header "AUTHORS"
Jaakko Koivuniemi
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIpipicfile\fR\|(1), \fIi2cdetect\fR\|(8), \fIi2cset\fR\|(8), \fIi2cget\fR\|(8)


12f675/pipic/pipic.cxx
/**************************************************************************
*
* Send commands to a PIC processor on i2c bus connected to Raspberry Pi and
* read data from the PIC.  
*       
* Copyright (C) 2013 Jaakko Koivuniemi.
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program. If not, see <http://www.gnu.org/licenses/>.
*
****************************************************************************
*
* Fri Jul 26 20:28:06 CEST 2013
* Edit: Sun Jul 28 19:34:53 CEST 2013
* 
* Jaakko Koivuniemi
**/

#include <iostream>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <linux/i2c-dev.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <getopt.h>
#include <errno.h>

void printusage()
{
  std::cout << "usage: pipic -a address [-c command [-d data] | -r b|w|W] [-h] [-v] [-V]\n";
}

void printversion()
{
  std::cout << "pipic v. 20130728, Jaakko Koivuniemi\n";
}


int main(int argc, char **argv)
{  
  int verb=0; // 1=verbosed output
  int rbyte=0; // read byte
  int rword=0; // read 16-bit word
  int rword2=0; // read 32-bit word
  int rdata=0; // read data
  int wbyte=0; // write byte
  int wword=0; // write 16-bit word
  int wword2=0; // write 32-bit word
  int wdata=0; // data to write
  int rnxt=0;
  int fd;
  char fileName[100]="/dev/i2c-1";
  int  address=0x00;
  int  cmd=-1; // no operation
  unsigned char buf[10];
  char nstr[100]="";

  int optch=0;
  while(optch!=-1)
    {
      optch=getopt(argc,argv,"a:c:d:r:hvV");
      if(optch=='a')
	{
	  address=atoi(optarg);
	}
      if(optch=='c')
	{
          cmd=atoi(optarg);
	}
      if(optch=='d')
        {
          if(optarg[strlen(optarg)-1]=='b') wbyte=1;
          if(optarg[strlen(optarg)-1]=='w') wword=1;
          if(optarg[strlen(optarg)-1]=='W') wword2=1;
          wdata=atoi(optarg);          
        }
      if(optch=='r')
	{
          if(strcmp(optarg,"b")==0) rbyte=1;
          if(strcmp(optarg,"w")==0) rword=1;
          if(strcmp(optarg,"W")==0) rword2=1;
	}
      if(optch=='v')
	{
	  verb=1;
	}
      if(optch=='h')
	{
	  printusage();
	  return 0;
	}
      if(optch=='V')
	{
	  printversion();
	  return 0;
	}
    }

  if((address<0x03)||(address>0x77)||(cmd>255))
    {
      printusage();
      return -1;
    }

// open port for reading and writing
  if(verb==1) std::cout << "Open " << fileName << "\n";
  if((fd = open(fileName, O_RDWR)) < 0) 
  {
     std::cout << "Failed to open i2c port: " << strerror(errno) << "\n";
     return -1;
  }

  sprintf(nstr,"0x%02x",address);
  if(verb==1) std::cout << "Chip address " << nstr << "\n";
  if(ioctl(fd, I2C_SLAVE, address) < 0) 
  {
     std::cout << "Unable to get bus access to talk to slave: " 
     << strerror(errno) << "\n";
     return -1;
  }

  if((cmd>=0)&&(cmd<=255))
  { 
     buf[0]=cmd;
     if(wbyte==1)
     { 
        buf[1]=wdata;
        if(verb==1) printf("Send 0x%02x%02x\n",buf[0],buf[1]);
        if((write(fd, buf, 2)) != 2) 
        {
           std::cout << "Error writing to i2c slave: " << strerror(errno) << "\n";
           return -1;
        }
     }
     else if(wword==1)
     { 
        buf[1]=(int)(wdata/256);
        buf[2]=wdata%256;
        if(verb==1) printf("Send 0x%02x%02x%02x\n",buf[0],buf[1],buf[2]); 
        if((write(fd, buf, 3)) != 3) 
        {
           std::cout << "Error writing to i2c slave: " << strerror(errno) << "\n";
           return -1;
        }
     }
     else if(wword2==1)
     { 
        buf[1]=(int)(wdata/16777216);
        rnxt=wdata%16777216;
        buf[2]=(int)(rnxt/65536);
        rnxt=rnxt%65536;
        buf[3]=(int)(rnxt/256);
        buf[4]=rnxt%256;
        if(verb==1) printf("Send 0x%02x%02x%02x%02x%02x\n",buf[0],buf[1],buf[2],buf[3],buf[4]);
        if((write(fd, buf, 5)) != 5) 
        {
           std::cout << "Error writing to i2c slave: " << strerror(errno) << "\n";
           return -1;
        }
     }
     else
     {
        if(verb==1) printf("Send 0x%02x\n",buf[0]);
        if((write(fd, buf, 1)) != 1) 
        {
           std::cout << "Error writing to i2c slave: " << strerror(errno) << "\n";
           return -1;
        }
     }
  }


  if(rbyte==1)
  {
     if(read(fd, buf,1)!=1) 
     {
        std::cout << "Unable to read from slave: " << strerror(errno) << "\n";
        return -1;
     }
     else 
     {
        if(verb==1) printf("Receive 0x%02x\n",buf[0]);
        rdata=buf[0];
        printf ("%d\n",rdata);
     }
  } 
  else if(rword==1)
  {
     if(read(fd, buf,2)!=2) 
     {
        std::cout << "Unable to read from slave: " << strerror(errno) << "\n";
        return -1;
     }
     else 
     {
        if(verb==1) printf("Receive 0x%02x%02x\n",buf[0],buf[1]); 
        rdata=256*buf[0]+buf[1];
        printf ("%d\n",rdata);
     }
  }
  else if(rword2==1)
  {
     if(read(fd, buf,4)!=4) 
     {
        std::cout << "Unable to read from slave: " << strerror(errno) << "\n";
        return -1;
     }
     else 
     {
        if(verb==1) printf("Receive 0x%02x%02x%02x%02x\n",buf[0],buf[1],buf[2],buf[3]);
        rdata=16777216*buf[0]+65536*buf[1]+256*buf[2]+buf[3];
        printf ("%d\n",rdata);
     }
  }

  return 0;
}
